### ISSUE TITLE
Scaling limits in SpecCluster

### ISSUE THREAD

**Issue by jacobtomlinson** [dask/distributed#3054]
Currently in dask-kubernetes it is possible to set the config `kubernetes.count.max` to a value and it will not scale beyond that point.

While working on the rewrite of dask-kubernetes to use `SpecCluster` I've not been able to find a way to do this. 

Is there a new max scaling config option that I'm missing or do we need to look at adding this feature?

**Comment by mrocklin** [comment]
Adding it as a feature here makes sense to me.  cc'ing @lesteve and @guillaumeeb because they have been active in the dask-jobqueue SpecCluster rewrite.  cc'ing @jcrist as well in case we want to apply this more broadly to non-SpecCluster classes.

Adding a check in `SpecCluster.scale` seems easy.  We would probably want to change the configuration name.  What should the configuration name be?  `distributed.deploy.limits.n` ?  Will we want to limit by cores or memory or cost in the future?  If so, we should probably make sure that our configuration names here can extend easily.

Alternatively, if you wanted to avoid this problem for dask-kubernetes, you could probably add a small `scale` implementation to KubeCluster

```python
def scale(self, n):
    n = min(n, dask.config.get("kubernetes.count.max"))  # or something similar
    return super().scale(n)
```

**Comment by jcrist** [comment]
When a user calls `scale` with a value > max, what happens? Does it scale to the max? Does it error? I'm slightly leaning towards error, the user may expect that many workers for their code to work, silently getting fewer may lead to problems.

Also, I have a slight preference to namespacing limits by cluster manager, rather than a global max. This would allow users to have different limits when working locally and on kubernetes, without having to have different configs.

**Comment by jacobtomlinson** [comment]
@mrocklin for `dask-kubernetes` I'm tempted to add the shim you suggested along with a deprecation warning pointing to the new implementation. 

@jcrist I agree that silently giving you a different number could lead tos ome hard to track down bugs on the users end. Perhaps a warning would be less aggressive than an error though?

I also agree that different cluster managers will probably have different limits. Do you propose something like `distributed.deploy.limits.kubernetes.n`?

**Comment by jcrist** [comment]
The reason I like an error is it will be noticed for both interactive and scripts (where a warning in the logs may go unnoticed).

And I was thinking more of a naming convention, but fully namespaced under the subproject's config. `kubernetes.limits.max-workers`. This keeps the config for each cluster manager together. I also would prefer a more explicit name than `n` (n what?).

**Comment by lesteve** [comment]
> Will we want to limit by cores or memory or cost in the future?

It would make sense indeed to be able to express the limit in terms of `cores` or `memory`, i.e. the same way we can do it with `scale` (e.g. `cluster.scale(cores=40)`).

**Comment by TomAugspurger** [comment]
FWIW, I think an error or logging a warning are OK.

In an async `.scale` context, do we expect to be able to determine whether a user has requested too many resources synchronously?

**Comment by mrocklin** [comment]
> When a user calls scale with a value > max, what happens? Does it scale to the max? Does it error? I'm slightly leaning towards error, the user may expect that many workers for their code to work, silently getting fewer may lead to problems.

Currently in dask-kubernetes we don't give them any feedback.

1.  User: `cluster.scale(1000)` "Please give me 1000 workers"
2.  Cluster manager: "Here are ten workers"

This has worked out ok so far.  I'm not opposed to changing it, but I think it does its job well today.  If we're going to change this behavior in scale then we also need to change it in `adapt`, which calls scale frequently.   We would want to make sure that we're not blasting the user with warnings.

For the short term, I recommend that we keep the current policy in dask-kubernetes and just implement it with a shim.  That should unblock things for @jacobtomlinson and avoid blocking on policy discussions.

**Comment by mrocklin** [comment]
> In an async .scale context, do we expect to be able to determine whether a user has requested too many resources synchronously?

Yes.  This is all known immediately on our side.

**Comment by mrocklin** [comment]
It's also worth noting that `scale` is always immediate, and never waits on anything.

**Comment by guillaumeeb** [comment]
> cc'ing @lesteve and @guillaumeeb because they have been active in the dask-jobqueue SpecCluster rewrite

I don't think this is an important feature in batch scheduler context as the batch scheduler already provides it. That being said, this won't hurt either, and providing a default on an HPC platform could be helpful for some user, but only if he is informed of the limitation somehow.

**Comment by jcrist** [comment]
In dask-gateway I went with a warning if `scale` exceeded a resource limit (see https://github.com/jcrist/dask-gateway/pull/93). I think that'd make sense here (and in the case of adaptive case this should be silenced or the `adapt` max trimmed to below the user limit).
